frontend/model/database.js

```diff
@@ -9,12 +9,42 @@ const log = localforage.createInstance({
   storeName: 'Contracts'
 })
 
+const isLightweightClient = process.env.LIGHTWEIGHT_CLIENT ?? true
+
 // make gi.log use localforage for storage
-sbp('sbp/selectors/overwrite', {
-  'gi.db/get': key => log.getItem(key),
-  // TODO: handle QuotaExceededError
-  'gi.db/set': (key, value) => log.setItem(key, value)
-})
+sbp('sbp/selectors/overwrite',
+  isLightweightClient
+    ? {
+        'gi.db/get': key => {
+          const contractId = sbp('gi.db/log/isLogHEAD', key)
+          if (!contractId) {
+            return null
+          }
+          const state = sbp('state/vuex/state')
+          if (state?.contracts?.[contractId]?.HEAD) {
+            return state.contracts[contractId].HEAD
+          }
+          return null
+        },
```

Comment:
So the point of `'gi.db/get'` is to return the value for a key. That doesn't seem to be what this is doing, and what it is doing is entirely unclear to me.

It seems to be some kind of weird version of `'gi.db/log/logHEAD'`...? Can you help me figure what you're  trying to do here and why?

Reply:
The implementation is a bit different from what we discussed earlier. I figured that storing messages just for deleting them afterwards was a bit redundant and added unnecessary complexity. Hence, with this change light clients will not store messages at all, only the contract state. So, this is returning `null` for normal message keys (as there is nothing to return) and returning the current HEAD for HEAD keys.

Reply:
@corrideat OK, is there a reason to return the head instead of null as well?

And can you create a helper function instead of defining a selector? (Selectors are meant to be used as an API... and I'm not sure we should add `isLogHEAD` as an API...)

Reply:
@taoeffect  Yes, there's a reason. The rest of the client logic expects the HEAD to have a valid value (note that with this implementation it's only stored in the VueX state).

I think that `gi.db/log/isLogHEAD` is as deserving of a selector as `gi.db/log/logHEAD` is. A helper function works, but I feel like it makes things a bit unbalanced, if you think of `gi.db/log/isLogHEAD` as the reverse operation of `gi.db/log/logHEAD`.

While answering this, I'm thinking of a cleaner implementation (but with more changes required), which is about how HEAD storage is handled. On the client side (talking about non-light clients or the current implementation without this PR), the HEADs are stored twice client-side. First, they are stored in the 'Contracts' store, and then they are stored in the VueX user store. This is some unnecessary duplication, and the client doesn't really need the HEADs in the 'Contracts' store (and having it is what actually caused issues with the first implementation that deleted messages).

The issue is that the server does need to store the HEAD, and that there is no server-side state the way that there is for a client/user. So, one way of making this a bit cleaner might be _not_ using `gi.db/get` and `gi.db/set` for HEADs, but using some specialised method (say we call them `gi.db/getHEAD` and `gi.db/setHEAD`). This way, we address several issues at once:

1. We no longer need `gi.db/log/logHEAD` nor `gi.db/log/isLogHEAD`
2. We can handle these operations differently in the client and in the server. The server could continue to append `-HEAD` to the keys used for storage. The client could for simply store the HEAD in VueX and (for non-light clients) store messages in the 'Contracts' store as it does now.
3. I believe it makes non-light more more robust (because the HEAD state is not unnecessarily duplicated)

Reply:
> I think that `gi.db/log/isLogHEAD` is as deserving of a selector as `gi.db/log/logHEAD` is.

Alright, we'll keep it.

> The issue is that the server does need to store the HEAD, and that there is no server-side state the way that there is for a client/user. So, one way of making this a bit cleaner might be not using gi.db/get and gi.db/set for HEADs, but using some specialised method (say we call them gi.db/getHEAD and gi.db/setHEAD). This way, we address several issues at once:

OK, these are some fair observations. Code duplication of any kind does get on my nerves of course... however, if we want to dedup things, we might want to use localstorage over Vuex in this instance, instead of the other way around.

The reason being is that when I designed this entire system, I design it with decentralized networks IPFS, Dat, et. al. in mind. I found the idea of being able to boot an entire app given a single hash (key) very appealing. The entire app could be represented and verified and loaded using this one hash with such an approach. The network that supported these key/value pairs, worked on key/value pairs. And so, the idea would be to do as much though a key/value API as possible. Moving things to specialized selectors like `getHEAD` and `setHEAD`, would potentially complicate any switch over to a decentralized backend. On the other hand, keeping everything built around a single key/value API would simplify any such migration.

That's the main reason. The other reason I'm hesitant to split things up into getHEAD/setHEAD (instead of just get/set), is that part of the reasoning here seems to be that we could make there be a bigger difference between the client and server. Well, that would increase the complexity of the system, because you'd have to think about the way the client and server behave differently, and we want to minimize the difference between them. The reason for this is, again, partly to support a fully decentralized "backend", but also, there was another potential vision, which was that anyone could trivially convert their client into a server if they wanted to. The idea was that at some point we'd release a desktop app version of GI, and if someone in the group decided that they want to be the server for the group, they could use the same client they were using, to run it as a server with a flip of a switch.

Let me know your thoughts given all of that.

Reply:
In this case, I think we can let `isLogHEAD` stay (I'll rename it to something more descriptive that explains why it returns a contract ID), address the issues with the env variable and call this issue closed. I have some ideas still for how to handle HEAD, but these are secondary to this issue and can be addressed separately.

Reply:
> they could use the same client they were using, to run it as a server with a flip of a switch.
That's a very interesting idea!